enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,634 @@
+    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
+}
+```
+
+```go
+// RouteHTTPHeaderAction is a policy for setting, replacing or appending HTTP headers.
+//
+// +kubebuilder:validation:Enum=Append;Replace
+type RouteHTTPHeaderAction string
+
+const (
+	// AppendHTTPHeaderPolicy appends the header, preserving any existing header.
+	AppendHTTPHeader RouteHTTPHeaderPolicy = "Append"
+	// ReplaceHTTPHeaderPolicy sets the header, removing any existing header.
+	ReplaceHTTPHeaderPolicy RouteHTTPHeaderPolicy = "Replace"
+)
+```
+
+The RouteDeleteHTTPHeader is an empty union member .
+```go
+// RouteDeleteHTTPHeader is used specify delete action per header.
+type RouteDeleteHTTPHeader struct {
+}
+```
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the
+frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of percent encoded values of
+the form `name:value` separated by comma.
+The template translates these values into the appropriate `http-response set-header` stanzas.
+For example, if `ROUTER_SET_HTTP_HEADERS` has percent encoded value `X-Frame-Options%3A+DENY,X-XSS-Protection%3A+1%3Bmode%3Dblock` 
+of `X-Frame-Options: DENY:Set,X-XSS-Protection: 1;mode=block` then the template adds the following stanzas to
+the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:
+```conf
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Deleting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_DELETE_HTTP_HEADERS`, which specify the headers to set in the
```

Comment:
I think you want to use the same environment variable for `spec.httpHeaders.headersManipulation.response[*].set` and `spec.httpHeaders.headersManipulation.response[*].delete` in order to preserve ordering.

Reply:
i think two env vars is fine. If we want only one then I will have to append Action to name:value:action in case of Set and name:action incase of Delete

Reply:
i think I get your point but then we will have pass Action as mentioned above in the env var.